crates/wysiwyg/src/composer_model/hyperlinks.rs

```diff
@@ -58,9 +61,63 @@ where
 
     pub fn set_link(&mut self, link: S) -> ComposerUpdate<S> {
         self.push_state_to_history();
-        let (s, e) = self.safe_selection();
-        let range = self.state.dom.find_range(s, e);
-        self.set_link_range(range, link)
+        let (mut s, mut e) = self.safe_selection();
+
+        let mut range = self.state.dom.find_range(s, e);
+
+        // Find container link that completely covers the range
+        if let Some(link) = self.find_parent_links(&range) {
+            // If found, update the range to the container link bounds
+            range = self.state.dom.find_range_by_node(&link);
+            (s, e) = (range.start(), range.end());
+        }
+
+        if s == e {
+            return ComposerUpdate::keep();
+        }
+
+        let inserted = self
+            .state
+            .dom
+            .insert_parent(&range, DomNode::new_link(link, vec![]));
+
+        // Ensure no duplication by deleting any links contained within the new link
+        self.delete_child_links(&inserted);
+
+        self.create_update_replace_all()
+    }
+
+    fn delete_child_links(&mut self, node_handle: &DomHandle) {
+        let node = self.state.dom.lookup_node(node_handle);
+
+        for handle in node
+            .iter_containers()
+            .filter(|n| matches!(n.kind(), ContainerNodeKind::Link(_)))
+            .map(|n| n.handle())
+            .filter(|h| *h != *node_handle)
+            .collect::<Vec<_>>()
+            .into_iter()
+            .rev()
```

Comment:
I wasn't sure how to make `DomNode.iter_containers()` return a double ended iterator, hence `collect()` followed by `rev()`

Reply:
I don't think there's a better way at the moment, to be honest. In the quotes PR I made some changes to the iterators, but I don't know if we could avoid collecting those items even with those changes.